### DISCUSSION TITLE
Error: Db is busy: Profile is already in use by another process

### DISCUSSION THREAD

**Discussion by spowell-216** [iTwin/itwinjs-core#5794] [Q&A] [answered]
we are getting the following error message on IModelHost.startup().

`Error: Db is busy: Profile [/Application/app/server/src/server/assets/temp/obj-builder/obj-builder-7fcfc978b9-6bg6m/cache/profiles/obj-builder-7fcfc978b9-6bg6m] is already in use by another process`

this is a backend only application.  Everything runs fine locally on Windows, but when we deploy to AKS is when we are seeing the error.  

everything after `/temp` is a mounted Azure File share.  
`obj-builder-7fcfc978b9-6bg6m` is os.hostname()

code snipet for starting up IModelHost

```
const imhOptions: IModelHostOptions = {
   appAssetsDir: '/Application/app/server/src/server/assets',
   cacheDir: '/Application/app/server/src/server/assets/temp/obj-builder/obj-builder-7fcfc978b9-6bg6m/cache',
   authorizationClient: valid AuthorizationClient,
   logTileLoadTimeThreshold: IModelHostConfiguration.defaultLogTileLoadTimeThreshold,
   logTileSizeThreshold: IModelHostConfiguration.defaultLogTileSizeThreshold,
   tileContentRequestTimeout: IModelHostConfiguration.defaultTileRequestTimeout,
   tileTreeRequestTimeout: IModelHostConfiguration.defaultTileRequestTimeout,
   profileName: os.hostname()
};

await IModelHost.startup(imhOptions);
IModelHost.applicationId = 'obj-builder';
IModelHost.applicationVersion = "v1.0";
```

we are using the `node:18-bullseye-slim` docker image.
and using version `^4.0.6` of the `@itwin` libraries

let me know if you require any additional information to help troubleshoot the issue.
Thanks.

**Comment by kabentley** [comment, marked as answer]
You cannot have more than one iTwin.js process running at the same time using the same profile directory. You should choose something other than `os.hostname()` to make it unique. See [this note](https://github.com/iTwin/itwinjs-core/blob/74de8990ce5eb66761e50a423fc5d418eac34597/core/backend/src/IModelHost.ts#L282-L290).

**Comment by pratikshan85** [comment]
If the profile is already in use, it means they are running multiple itwin.js backend processes on the same machine. They must specify a separate profile ([as documented here](https://www.itwinjs.org/reference/core-backend/imodelhost/imodelhost/profilename/)). Likely they don't want multiple processes, in which case they need to find out why they're accidentally spawning multiple.

**Comment by spowell-216** [comment]
@kabentley @pratikshan85 thanks for the quick reply.  we did see that note for profile and why is we implemented the os.hostname()  to get the Kubernetes pod name so we have ensured that they are unique folder path between the different instances.

**Comment by kabentley** [comment]
somehow you must be starting more than one process per pod. 

BTW, if the `cacheDir` is already unique per process (probably a good idea, depending on what you're application does) you don't need to worry about `profileName`, since all that does is create a subdirectory below `cacheDir`. But the error you're seeing can only happen if there are two active processes using the same profile directory.

**Reply by spowell-216** [reply]
agreed, but we've been adding more and more unique values to try and troubleshoot will clean up once we figure out the issue. 

to rule out `os.hostname()` we tried again with `profile: Guid.createValue()` resulting in the same error : 

Error: Db is busy: Profile [/Application/app/server/src/server/assets/temp/obj-builder/`obj-builder-7cd6fbd795-nhgg2`/cache/profiles/`80eb0f0f-5965-4c60-a259-c627dd77ecdc`] is already in use by another process

we turned on Logger.Trace, and this is the last successful message before throwing the error above:
`Trace | CloudSqlite | 14:32:58.530 | create CloudCache`  

we think it's something in this call [this.appWorkspace.getCloudCache()](https://github.com/iTwin/itwinjs-core/blob/74de8990ce5eb66761e50a423fc5d418eac34597/core/backend/src/IModelHost.ts#L430)

**Reply by kabentley** [reply]
Hmmm, very strange. The (sole) purpose of the call to `this.appWorkspace.getCloudCache()` is to detect whether there's another process running with the same profile.

The `Db is busy` message indicates that we're trying to make 2 connections to the same cloud cache. Since it happens with a unique GUID we can assume there's only one process, so something else is failing. You don't see two messages in the log with "create CloudCache" do you?

**Reply by spowell-216** [reply]
we have verified we are only attempting to start one instance of `IModelHost`.  we have also set our config to only spin up 1 pod of our app, still getting same error.  

another thought, but I can't prove it.  Since we have mounted this `cacheDir` to an `Azure File Share (AFS)`, and `fs.mkdirSync()` is being used to create the additional folders.  I think there's a lag on AFS's side before the folders are available and ready for use.

**Reply by spowell-216** [reply]
no multiple error messages, just the single one being passed back through the try/catches

**Reply by kabentley** [reply]
Ahhh, that's almost undoubtedly the problem - I should have asked you about that. Your cacheDir should be on a local drive. Network files simply don't work well for SQLite, and even if they do work performance will be terrible. You can either use the OS dir or an ephemeral drive. Do you have one?

**Reply by kabentley** [reply]
see: https://www.sqlite.org/useovernet.html

**Reply by spowell-216** [reply]
in the past, we have been successful with the local pod temp dir, but it seems in recent iTwin versions, there a cacheSize requirement that exceeds our pod size

**Reply by kabentley** [reply]
I'm sorry to hear that. AFS simply isn't an appropriate solution for a cacheDir. We should try to figure out what's using space in the cacheDir. How large are your tmp drives?

**Reply by spowell-216** [reply]
I agree AFS or any remote storage should not be used. Is there any way we can set the cachDir size in `IModelHostOptions` that can be passed along to [makeCache](https://github.com/iTwin/itwinjs-core/blob/74de8990ce5eb66761e50a423fc5d418eac34597/core/backend/src/CloudSqlite.ts#L662).  the default `'10G'` is too big for our  Kubernetes pod.

**Reply by kabentley** [reply]
We will have to add that. I don't know why that's not there now. It won't use 10G anyway now, since you don't have any workspace data to load.

In the meantime, perhaps if you create the profile directory in your code and put a (1s?) wait before calling startup. Maybe that will cause the "busy" error from the AFS "drive" to go away.

**Reply by kabentley** [reply]
Also, can you try an experiment on your Windows machine where things work? Run your program and do a "dir/s" on the cacheDir after it finishes so we can see what's using the disk space. My guess is its the briefcase cache.

**Reply by spowell-216** [reply]
thanks for the suggestion, we will try that recommendation in the meantime.
